okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java

```diff
@@ -443,18 +474,37 @@ private void relayToResponse(HttpServerResponse hres,
    */
   private void authResponse(HttpClientResponse res, ProxyContext pc) {
     String modTok = res.headers().get(XOkapiHeaders.MODULE_TOKENS);
+    HttpServerRequest req = pc.getCtx().request();
     if (modTok != null && !modTok.isEmpty()) {
       JsonObject jo = new JsonObject(modTok);
       for (ModuleInstance mi : pc.getModList()) {
         String id = mi.getModuleDescriptor().getId();
+        String pathPattern = mi.getRoutingEntry().getPathPattern();
+
         if (jo.containsKey(id)) {
           String tok = jo.getString(id);
           mi.setAuthToken(tok);
           pc.debug("authResponse: token for " + id + ": " + tok);
+
+          tokenCache.put(pc.getTenant(),
+              req.method().name(),
+              pathPattern == null ? req.path() : pathPattern,
+              res.getHeader(XOkapiHeaders.USER_ID),
+              res.getHeader(XOkapiHeaders.PERMISSIONS),
+              req.getHeader(XOkapiHeaders.TOKEN),
+              tok);
         } else if (jo.containsKey("_")) {
           String tok = jo.getString("_");
           mi.setAuthToken(tok);
           pc.debug("authResponse: Default (_) token for " + id + ": " + tok);
+
+          tokenCache.put(pc.getTenant(),
+              req.method().name(),
+              pathPattern == null ? req.path() : pathPattern,
+              res.getHeader(XOkapiHeaders.USER_ID),
+              res.getHeader(XOkapiHeaders.PERMISSIONS),
+              req.getHeader(XOkapiHeaders.TOKEN),
+              tok);
```

Comment:
duplicate code can be extracted:

    } else {
      continue;
    }
    mi.setAuthToken(tok);
    tokenCache.put(...